source/inputoutput.rst

```diff
@@ -0,0 +1,114 @@
+Inputs and Outputs in BharatSim
+===============================
+
+Inputs
+------
+**to be completed**
+
+* function should feed in the csv file 1 row at a time
+* data gets 'ingested' and graphNodes are created
+
+Outputs
+-------
+
+A convenient way to store the output is by using a `CSV <https://en.wikipedia.org/wiki/Comma-separated_values>`_ file. Scala is `capable of writing to files <https://alvinalexander.com/scala/how-to-write-text-files-in-scala-printwriter-filewriter/>`_, but BharatSim simplifies the process when it comes to CSV outputs.
+
+.. note:: In case the quantities you'd like to output are fairly simple, you could use Scala's ``println`` function to directly output what you need.
+
+Saving your output to a CSV file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+BharatSim relies on a trait called ``SimulationListener`` to help output data .
+
+``SimulationListener`` contains 4 methods, each of which allow us to perform a task in one of the following situations:
+
+* At the start of the simulation
+* At the start of every time step
+* At the end of every time step
+* At the end of the simulation
+
+The BharatSim engine also contains a class called ``CsvOutputGenerator``, an extension of ``SimulationListener`` which has two attributes:
+
+* ``path``, the desired path for the output file to be stored
+* ``csvSpecs``, a user-defined class that outputs the headers and the rows required. Note that this user-defined class should extend the ``CSVSpecs`` trait and override the ``getHeaders`` and ``getRows`` methods.
+
+This class writes the headers at the start of the simulation, writes the rows at the start of every time step, and closes the writer at the end of the simulation.
+
+Output at a single instant of time
```

Comment:
This sounds a little confusing, but I can't come up with a better title yet. I'll keep this comment open so that I remember.